### PR TITLE
fix #69: deterministic document staleness via doc CLI subcommands

### LINKED ISSUE #69: Deterministic document staleness detection via doc CLI subcommands
## Description

Replace the current text-marker-based staleness detection (`⚠️ **STALE**` blockquotes scanned by the LLM) with deterministic CLI subcommands that read and write YAML frontmatter. This makes staleness a structured tool-call operation instead of unreliable text parsing.

### Problem

The current system requires skills to instruct the LLM to scan the first 10 lines of pipeline artifacts for a `⚠️ **STALE**` marker. This is unreliable because:
- LLMs don't always execute the "check for staleness" instruction
- Text scanning produces inconsistent results depending on context pressure
- The result is an implicit judgment, not a structured machine-readable signal

### Solution

Three picocli subcommands under a `doc` parent command:

```
camel kit doc check <file>                              # Query staleness → JSON
camel kit doc stale --reason "..." [--cascade] <file>    # Mark stale
camel kit doc unstale <file>                             # Clear staleness
```

### Frontmatter Schema

Each pipeline artifact is created with YAML frontmatter containing two namespaced blocks:

```yaml
---
staleness:
  stale: false
  since: null
  reason: null
generated:
  at: "2026-05-13T09:00:00Z"
  by: camel-plan
  from: design-spec.md
---
```

- `staleness` — mutable, written by `doc stale` and `doc unstale`
- `generated` — immutable after creation, written by the pipeline skill that produces the artifact
- `generated.from` enables data-driven cascade without hardcoded pipeline topology

### `doc check` Output

Returns JSON to stdout:

```json
{
  "file": "docs/camel-kit/007/implementation-plan.md",
  "stale": true,
  "since": "2026-05-13T10:00:00Z",
  "reason": "design-spec.md was refined",
  "generated": {
    "at": "2026-05-13T09:00:00Z",
    "by": "camel-plan",
    "from": "design-spec.md"
  }
}
```

Exit code 0 for successful execution regardless of staleness. Non-zero for errors only. Documents without frontmatter return `{"stale": false, ...null fields}`.

### Cascade

`--cascade` flag on `doc stale` walks the `generated.from` chain in sibling files to propagate staleness to all downstream artifacts. Single-file marking is the default.

### Skill Contract

- Each pipeline skill writes frontmatter when generating artifacts (documents are "born with metadata")
- Skills call `doc check` via tool call instead of scanning text
- `--reason` is required on `doc stale` for audit trail
- `⚠️ **STALE**` text markers are removed entirely — frontmatter is the sole source of truth

### Implementation

```
command/doc/
  DocCommand.java          # @Command parent (mirrors GraphCommand)
  DocCheckCommand.java     # Reads frontmatter, outputs JSON
  DocStaleCommand.java     # Sets staleness fields
  DocUnstaleCommand.java   # Clears staleness fields
```

### Tasks

- [ ] Implement `DocCommand` parent and three subcommands (`check`, `stale`, `unstale`)
- [ ] YAML frontmatter parser (extract/modify/write between `---` delimiters)
- [ ] `--cascade` flag with `generated.from` chain walking
- [ ] JSON output for `doc check`
- [ ] Update pipeline skill files to write frontmatter on artifact creation
- [ ] Update pipeline skill files to call `doc check` instead of scanning for `⚠️ **STALE**`
- [ ] Update `pipeline-infrastructure.md` to describe the new staleness system
- [ ] Update documentation in `docs/` to reflect the new `doc` commands

### Related

- Supersedes the staleness marker section of #65

### PR DESCRIPTION
**Pull request by luigidemasi**
## Summary

- Adds three picocli subcommands under a `doc` parent command: `doc check`, `doc stale`, `doc unstale`
- Adds `doc init --by <skill> --from <source> <file>` command for deterministic frontmatter metadata
- Implements `FrontmatterHandler` utility for YAML frontmatter parsing/modification using `jackson-dataformat-yaml`
- `doc check` reads frontmatter and outputs structured JSON (exit code 0 regardless of staleness)
- `doc stale --reason "..." [--cascade]` marks files stale with optional downstream propagation via `generated.from` chain
- `doc unstale` clears staleness from a document
- `doc init` adds provenance metadata (generated.at/by/from) — idempotent, preserves existing frontmatter
- Updates `pipeline-infrastructure.md` to document the new frontmatter schema and CLI commands
- Updates all pipeline skills (`camel-plan`, `camel-execute`, `camel-validate`, `camel-ship`) to use `doc check` / `doc stale` / `doc init` instead of scanning for `⚠️ **STALE**` text markers
- Adds frontmatter examples to artifact templates in `camel-plan` and `camel-validate` SKILL.md files
- Adds `camel-kit doc` section to `docs/commands.md`, updates `README.md`, `docs/architecture.md`, and `docs/user-guide.md`

## Review feedback addressed

- Added blank `--reason` guard (rejects `--reason ""`)
- Added CRLF support (`\r\n` normalized to `\n` before parsing) with 12 CRLF-specific tests
- Fixed cascade target in skill files — regenerated artifacts are not marked stale, only downstream ones
- Clarified staleness contract in `pipeline-infrastructure.md` (amendments vs regenerations)
- Scoped staleness checks in `state-management.md` to 5 canonical pipeline artifacts
- Declined `GeneratedInfo` immutability — setters needed for Jackson `convertValue()` deserialization
- Fixed cascade ambiguity in `camel-execute/SKILL.md` and `camel-plan/SKILL.md` — removed "for each" loop language, clarified single `--cascade` call
- Added `--cascade` flag to `camel-validate/SKILL.md` for consistency with other pipeline skills
- Added `bash` language identifiers to fenced command blocks in `pipeline-infrastructure.md`
- Broadened exit code 1 description in `commands.md` to include argument validation errors
- Added missing `<file>` argument to `doc stale` example in `user-guide.md`
- Added frontmatter examples to artifact templates + `doc init` CLI fallback to fix the "born without metadata" cascade chain gap

## Trait coverage gaps addressed

Consistency audit of all pipeline skills and agent traits identified coverage gaps:

- **bob/camel-plan**: Returns control to mode system orchestrator instead of auto-invoking `camel-execute` (plan should produce its artifact and return — pipeline orchestration belongs to `camel-ship`)
- **opencode/camel-plan**: Returns control to orchestrator with step budget guidance
- **opencode/camel-brainstorm**: LLM-parallel tool calls, `Plan` agent type selection, `Explore` for migration scanning, LSP opt-in
- **qwen/camel-brainstorm**: Fork-based research isolation, `todo_write` progress tracking, named subagent for architecture analysis
- **claude/camel-validate**: `TaskCreate` progress tracking, `ScheduleWakeup` build pacing, parallel dimension analysis via `Agent` tool

## Test plan

- [x] 34 unit tests for `FrontmatterHandler` (parsing, modification, round-trip, edge cases, 12 CRLF tests)
- [x] 16 CLI command tests for `DocCheckCommand`, `DocInitCommand`, `DocStaleCommand`, `DocUnstaleCommand` (including cascade, blank reason, init idempotency)
- [x] Full reactor build passes (346 tests total, 0 failures)
- [x] Verified `applyTraits()` skill names list includes all 5 new trait targets

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/command/doc/DocStaleCommand.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+@Command(name = "stale", description = "Mark a document as stale")
+public class DocStaleCommand implements Callable<Integer> {
+
+    @Parameters(index = "0", description = "Path to the document to mark stale")
+    Path file;
+
+    @Option(names = "--reason", required = true, description = "Reason for staleness")
+    String reason;
+
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/command/doc/DocStaleCommand.java:31]: Fixed — added `reason.isBlank()` guard before `markFileStale()`. Also added a test (`staleRejectsBlankReason`) that verifies blank reason returns exit code 1 and leaves the file untouched.

File: camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/doc/FrontmatterHandler.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        return content != null && content.startsWith(DELIMITER);
+    }
+
+    public static String extractFrontmatterYaml(String content) {
+        if (!hasFrontmatter(content)) {
+            return null;
+        }
+        int start = DELIMITER.length();
+        int end = content.indexOf("\n" + DELIMITER, start);
+        if (end < 0) {
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/doc/FrontmatterHandler.java:38]: Fixed — added `normalizeCrlf()` method that converts `\r\n` to `\n` at the start of `extractFrontmatterYaml()` and `extractBody()`. Also added 5 CRLF-specific tests (`crlfHasFrontmatter`, `crlfExtractFrontmatterYaml`, `crlfParseStaleness`, `crlfExtractBody`, `crlfToCheckJson`).

File: camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/doc/GeneratedInfo.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    public void setBy(String by) {
+        this.by = by;
+    }
+
+    public String getFrom() {
+        return from;
+    }
+
+    public void setFrom(String from) {
+        this.from = from;
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/java/io/github/luigidemasi/camelkit/doc/GeneratedInfo.java:39]: Declining this suggestion. `GeneratedInfo` needs setters for Jackson's `convertValue()` deserialization from the YAML frontmatter map. Making it immutable would require adding `@JsonCreator` annotations — added complexity for no practical gain, since the class is only constructed internally by `FrontmatterHandler`. The setters are part of the standard Jackson POJO contract.

File: camel-kit-core/src/main/resources/skills/camel-plan/SKILL.md [hunk 4]
```diff
@@ -225,7 +225,7 @@ Fix any issues inline.
 After saving the plan:
 
 1. Save the implementation plan to `docs/camel-kit/<PIPELINE_ID>/implementation-plan.md`
-2. **Mark downstream artifacts stale** — per `shared/pipeline-infrastructure.md`, check for existing downstream artifacts (`execution-report.md`, `validation-report.md`, `stamp-report.md`) and prepend the staleness marker if they exist
+2. **Mark downstream artifacts stale** — per `shared/pipeline-infrastructure.md`, run `{COMMAND_PREFIX} doc stale --reason "implementation plan regenerated" --cascade <implementation-plan.md>` to propagate staleness to downstream artifacts
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/resources/skills/camel-plan/SKILL.md]: Fixed — updated the instruction to target downstream artifacts only, not the freshly regenerated plan itself. The same fix was applied to `camel-execute/SKILL.md` and `camel-validate/SKILL.md` for consistency.

File: camel-kit-core/src/main/resources/skills/camel-ship/guides/state-management.md [hunk 5]
```diff
@@ -123,8 +123,8 @@ When `--resume` is specified, the state management layer adds a staleness check
 
 1. Read `.camel-kit/pipeline.json`
 2. Verify `mode` is `"ship"`
-3. Scan all artifacts in `docs/camel-kit/<activePipeline>/` for staleness markers
-4. If any artifact has `⚠️ **STALE**` in its first 10 lines:
+3. For each artifact in `docs/camel-kit/<activePipeline>/`, run `{COMMAND_PREFIX} doc check <artifact>` and parse the JSON output
+4. If any artifact has `"stale": true` in the JSON response:
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/resources/skills/camel-ship/guides/state-management.md]: Fixed — scoped the staleness check to the 5 canonical pipeline artifacts (`design-spec.md`, `implementation-plan.md`, `execution-report.md`, `validation-report.md`, `stamp-report.md`) instead of iterating all files in the directory.

File: camel-kit-core/src/main/resources/skills/camel-execute/SKILL.md [hunk 6]
```diff
@@ -403,7 +403,7 @@ Verification: PASS/PARTIAL/FAIL/NOT_RUN
 
 Save the completion summary as `docs/camel-kit/<PIPELINE_ID>/execution-report.md`.
 
-**Mark downstream artifacts stale** — per `shared/pipeline-infrastructure.md`, check for existing downstream artifacts (`validation-report.md`, `stamp-report.md`) and prepend the staleness marker if they exist.
+**Mark downstream artifacts stale** — per `shared/pipeline-infrastructure.md`, for each existing downstream artifact (`validation-report.md`, `stamp-report.md`), run `{COMMAND_PREFIX} doc stale --reason "execution report regenerated" --cascade <first-downstream-artifact>` to propagate staleness. Do NOT mark the freshly regenerated `execution-report.md` itself stale.
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/resources/skills/camel-execute/SKILL.md]: Fixed — removed the "for each" loop language that caused duplicate cascade writes. The instruction now says to run `--cascade` once from the first downstream artifact and explicitly warns "do NOT loop over individual artifacts." Applied the same fix to `camel-plan/SKILL.md` which had identical ambiguity.

File: camel-kit-core/src/main/resources/skills/shared/pipeline-infrastructure.md [hunk 7]
```diff
@@ -206,58 +206,91 @@ This enables iterative refinement without restarting the entire pipeline.
...
-|---|---|
-| `implementation-plan.md` | Yes |
-| `execution-report.md` | Yes |
-| `validation-report.md` | Yes |
-| `stamp-report.md` | Yes |
+```
+{COMMAND_PREFIX} doc check <file>                              # Query staleness → JSON
+{COMMAND_PREFIX} doc stale --reason "..." [--cascade] <file>   # Mark stale
+{COMMAND_PREFIX} doc unstale <file>                            # Clear staleness
+```
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply camel-kit-core/src/main/resources/skills/shared/pipeline-infrastructure.md]: Fixed — added `bash` language identifiers to both fenced command blocks (lines 236 and 284). These are Qute templates where `{COMMAND_PREFIX}` gets substituted before the agent sees them, so `bash` is the correct language after substitution.

File: docs/commands.md [hunk 8]
```diff
@@ -337,6 +339,94 @@ $ camel-kit nextId inventory-sync
...
+camel-kit doc unstale docs/camel-kit/001-order-processing/implementation-plan.md
+```
+
+**Exit Codes:**
+
+| Code | Meaning |
+|---|---|
+| 0 | Success — command completed (for `check`: regardless of staleness) |
+| 1 | Error — file not found or I/O failure |
+
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply docs/commands.md:442]: Fixed — broadened exit code 1 description to: "Error — file not found, I/O failure, or invalid arguments (e.g., blank `--reason`)". This covers the blank `--reason` guard added in the previous review round.

File: docs/user-guide.md [hunk 9]
```diff
@@ -289,19 +289,19 @@ If you need to change a design spec after downstream artifacts have been generat
 When brainstorm detects that `design-spec.md` already exists, it enters **amend mode**:
 1. Loads the existing spec and presents it for modification
 2. After you approve the amendments, overwrites the design spec
-3. Marks all downstream artifacts (`implementation-plan.md`, `execution-report.md`, etc.) as **stale**
+3. Marks all downstream artifacts as **stale** via `camel-kit doc stale --reason "design spec was amended" --cascade`
```
[the remark this answers is not in the document]
Comment (luigidemasi) [reply docs/user-guide.md]: Fixed — added the missing `<file>` argument: `camel-kit doc stale --reason "design spec was amended" --cascade docs/camel-kit/001-order-processing/implementation-plan.md`.

**Comment by luigidemasi** [comment]
### Addressing remaining round-1 review items (from review body)

**Outside-diff: Staleness contract contradiction (pipeline-infrastructure.md:282-296)**
Addressed in commit 7e85df2 — clarified the two distinct flows:
- **Amendments** (partial edits): call `doc stale --cascade` to mark *downstream* artifacts stale. The amended artifact itself is NOT marked stale because it was just updated.
- **Regenerations** (full re-run): the regenerated artifact is written with `stale: false` frontmatter. It's born fresh.

The section heading was renamed from "When a skill writes or amends" to "Applying Staleness — Amendments" to eliminate the ambiguity. The separate "Regenerations" subsection (lines 290-296) clarifies the fresh-write contract.

**Nitpick: CRLF test coverage (FrontmatterHandlerTest.java)**
Addressed in commit 7e85df2 — added 5 CRLF-specific tests:
- `crlfHasFrontmatter` — detection with `\r\n` line endings
- `crlfExtractFrontmatterYaml` — YAML extraction from CRLF content
- `crlfParseStaleness` — staleness parsing from CRLF frontmatter
- `crlfExtractBody` — body extraction after CRLF frontmatter
- `crlfToCheckJson` — full JSON output from CRLF documents

All 5 use CRLF variants of the existing LF fixtures and assert identical behavior.

**Comment by luigidemasi** [comment]
### Addressing round-1 outside-diff comment: frontmatter in artifact templates

**Comment:** "Agents prioritize visual templates over procedural rules. If a template starts with `# [Project Name] Implementation Plan`, the agent is 90% likely to omit the YAML frontmatter."

**Analysis:** The reviewer is correct that the cascade chain breaks when artifacts lack frontmatter — `cascadeStale()` checks `gen != null` at `DocStaleCommand.java:98` and silently skips files without `generated.from`.

**Fix (two-layer):**

1. **Templates** — Added frontmatter examples to `camel-plan/SKILL.md` and `camel-validate/SKILL.md` templates so LLMs pattern-match the correct output structure (the reviewer's suggestion).

2. **`doc init` CLI command** — New command `doc init --by <skill> --from <source> <file>` as a deterministic fallback. Each skill's save instruction now includes a `doc init` call that guarantees correct frontmatter with real `Instant.now()` timestamps and valid YAML, regardless of whether the LLM included frontmatter in its output.

`doc init` is idempotent — if frontmatter already exists (because the LLM followed the template), it's preserved. The two layers don't conflict.

Why `doc init` in addition to templates: templates rely on LLM-generated YAML (risk of syntax errors, placeholder timestamps, wrong `from:` values). The CLI command produces deterministic output with guaranteed-valid YAML, real timestamps, and explicit provenance parameters.

**Commit:** 2a5d9a99 — 4 new tests, all 346 tests pass.
